File: Data Structures & Algorithms/binary-tree-right-side-view/submission-5.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number[]}
     */
    rightSideView(root) {
        /**
         * @param {number[]} acc
         * @param {TreeNode} tree
         * @param {number} l
         * @param {number} ml
         */
        const dfs = (acc, tree, l, ml) => {
            if (tree == null) return [acc, ml]
            console.log({ l, acc, ml })

            const left = dfs(l > ml ? [...acc, tree.val] : acc, tree.right, l + 1, Math.max(l, ml))
            const right = dfs(l > left[1] ? [...left[0], tree.val] : left[0], tree.left, l + 1, Math.max(l, left[1]))
            console.log({ left, right })
            return right
        }

        console.log(root)

        const [res, _] = dfs([], root, 0, -1)
        return res
    }
}

File: Data Structures & Algorithms/house-robber/submission-5.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        cache = {}
        def dfs(i: int) -> int:
            if i in cache:
                return cache[i]
            if i >= len(nums):
                return 0
            v = nums[i] + max(dfs(i + 2), dfs(i + 3))
            cache[i] = v
            return v
        
        return max(dfs(0), dfs(1))
        